### PR TITLE
Automated testing for exec docs using IE 

### PR DESCRIPTION
**Pull request by naman-msft**
Attached github action and its instructions for automated testing of exec docs

### CODE REVIEW

File: examples/test-exec-doc.yml
Comment (vmarcella) [inline examples/test-exec-doc.yml:41]: Could you pin this to a specific version?

Comment (naman-msft) [reply examples/test-exec-doc.yml:41]: the action works with this line. I tested it

File: examples/test-exec-doc.yml
Comment (vmarcella) [inline examples/test-exec-doc.yml:29]: Do we need to add the fetch-depth? Is this actually needed?

File: examples/test-exec-doc.yml
Comment (vmarcella) [inline examples/test-exec-doc.yml:68]: I would pin this to a specific version of IE, there's no guarantee that this example might work for future versions.

File: examples/test-exec-doc.yml
Comment (vmarcella) [inline examples/test-exec-doc.yml:32]: Can we use v2 for the azure login action?

File: examples/test-exec-doc.yml
Comment (vmarcella) [inline examples/test-exec-doc.yml:16]: Does triggering on any pull request with changes to markdown files create another pull request? Does this mean the pipeline will run twice for every push/pull request that gets created with markdown changes?

File: examples/test-exec-doc.yml
Comment (vmarcella) [inline examples/test-exec-doc.yml:104]: Shouldn't this only test files that are part of the most recent changes? What if a user has multiple markdown files in the same directory and none of them have changed?

File: examples/test-exec-doc.yml
Comment (vmarcella) [inline examples/test-exec-doc.yml:136]: I think you left your alias here as the assignee

File: examples/test-exec-doc.yml
Comment (vmarcella) [inline examples/test-exec-doc.yml:46]: Pin these to whatever version you're using during testing

File: examples/test-exec-doc.yml
Comment (vmarcella) [inline examples/test-exec-doc.yml:50]: Where does PAT come from? It's not mentioned in the documentation

Comment (vmarcella) [reply examples/test-exec-doc.yml:50]: My bad, I actually see that it does now.

File: examples/test-exec-doc.yml
Comment (vmarcella) [inline examples/test-exec-doc.yml:136]: Left yourself as the assignee

File: examples/test-exec-doc.yml
Comment (vmarcella) [inline examples/test-exec-doc.yml:96]: Shouldn't the branch that's being used for the push/PR be used for creating a new pull request? I'm not sure how the PR creation is working here

File: examples/test-instructions.md
Comment (vmarcella) [inline examples/test-instructions.md:52]: I would summarize only what the script portion of the pipeline does. The documentation about checking out, signing into azure, etc are all pretty self explanatory

File: examples/test-instructions.md
Comment (vmarcella) [inline examples/test-instructions.md:5]: How does this action prevent pull requests from merging?

**Review by vmarcella — CHANGES_REQUESTED** [review]
Left some comments about what needs to be changed before it can be merged in, but will review it again after everything has been resolved.

File: examples/test-exec-doc.yml
Comment (naman-msft) [reply examples/test-exec-doc.yml:29]: good point - based on GH copilot it isn't so have removed it. will test and see if it breaks at any point while executing

File: examples/test-exec-doc.yml
Comment (naman-msft) [reply examples/test-exec-doc.yml:68]: can you deploy the release with the spinner not showing up in the gh action? Currently, I have it pinned to v0.2.3 and will update it once you have the latest version installed

File: examples/test-exec-doc.yml
Comment (naman-msft) [reply examples/test-exec-doc.yml:16]: good one - it is triggering on both and we primarily want it to be triggered on changes to markdown files and that too only in the branch from where it was triggered. removed the pull_request bit

File: examples/test-exec-doc.yml
Comment (vmarcella) [reply examples/test-exec-doc.yml:68]: This work hasn't been done yet, but `v0.3.0` won't come out for a little bit because I'm including a v1 for pre-requisites in that release.

File: examples/test-exec-doc.yml
Comment (naman-msft) [reply examples/test-exec-doc.yml:104]: good one - made the change. it will only check the files that were updated as of the last commit. let me test it now

File: examples/test-instructions.md
Comment (naman-msft) [reply examples/test-instructions.md:5]: technically, other settings will need to be configured for the prevention step but it does show errors in docs as part of the PR. Now, someone could theoretically still merge the PR. updated the verbiage as a result

File: examples/test-exec-doc.yml
Comment (naman-msft) [reply examples/test-exec-doc.yml:96]: it works from the same branch

File: examples/test-exec-doc.yml
Comment (vmarcella) [reply examples/test-exec-doc.yml:41]: Yeah, I believe this uses the latest python and if there are newer versions which are incompatible either with the script or the packages that you use in the script, it will break. It's generally always better to pin to an exact version of what you're going to use for reproducibility as well.

File: examples/test-exec-doc.yml
Comment (naman-msft) [reply examples/test-exec-doc.yml:68]: so how would you remove the spinner for ignite?

Comment (vmarcella) [reply examples/test-exec-doc.yml]: I won't, it's too late of an ask

File: examples/test-exec-doc.yml
Comment (vmarcella) [inline examples/test-exec-doc.yml:26]: I thought we agreed to get rid of this?

File: examples/test-exec-doc.yml
Comment (vmarcella) [inline examples/test-exec-doc.yml:32]: Isn't the only scenario where this action could run on an initial commit is if the repository is initialized with both markdown files and this action? How likely is that and is this check needed?

File: examples/test-exec-doc.yml
Comment (vmarcella) [inline examples/test-exec-doc.yml:100]: I don't think you need to add fall backs here; It's highly unlikely that the pinned versions will fail installation because once they're published, they're effectively available for forever

Comment (vmarcella) [reply examples/test-exec-doc.yml:100]: Same for the line below

File: examples/test-exec-doc.yml
Comment (vmarcella) [inline examples/test-exec-doc.yml:204]: You're still assigning these issues to yourself here

File: examples/test-exec-doc.yml
Comment (vmarcella) [inline examples/test-exec-doc.yml:15]: So correct me if I'm wrong, but won't this action run on any push to any branch in their repository? I.E. If I push markdown to `vmarcella/testing` that has markdown file changes, won't this action run and create the pull request for me (Regardless of if the branch is ready for a pull request or not)?

File: examples/test-exec-doc.yml
Comment (vmarcella) [inline examples/test-exec-doc.yml:58]: Why is this logic only needed where this is the first commit, but not needed in cases where this is the nth commit? Isn't it possible for people to use branches other than main in pre-existing repositories?

File: examples/test-exec-doc.yml
Comment (vmarcella) [inline examples/test-exec-doc.yml:95]: Please pin this to python 3.12, there's no need to have this pipeline always using the latest python

File: examples/test-exec-doc.yml
Comment (vmarcella) [inline examples/test-exec-doc.yml:164]: If there are no changed files, I don't think the pipeline should reach this point, right?

File: examples/test-instructions.md
Comment (vmarcella) [inline examples/test-instructions.md:45]: I think it iterates over all *changed* files now, right?

**Review by vmarcella — CHANGES_REQUESTED** [review]
Still needs a few more changes, but it's getting closer to being ready for approval

File: examples/test-exec-doc.yml
Comment (naman-msft) [reply examples/test-exec-doc.yml:32]: it would run on an initial commit if a new branch is created and the first commit is made. in that case, it checks for changed files by comparing it to the latest commit of its upstream branch so if no changes are made from that last commit, it skips over this step. if there are changes, it tracks only those docs

File: examples/test-exec-doc.yml
Comment (naman-msft) [reply examples/test-exec-doc.yml:100]: what do you suggest? I put that because I was seeing some errors in installing those versions in the GH action shell and it seemed to fall back on the latest version. But maybe I was looking at the wrong thing. Just for posterity I added it

File: examples/test-exec-doc.yml
Comment (naman-msft) [reply examples/test-exec-doc.yml:26]: had to add it back because there were some issues while tracking changed filed between the previous commit and the current one - let's keep it

File: examples/test-exec-doc.yml
Comment (naman-msft) [reply examples/test-exec-doc.yml:204]: updated

File: examples/test-exec-doc.yml
Comment (naman-msft) [reply examples/test-exec-doc.yml:15]: i didn't understand - this action is triggered from the branch where markdown changes are made. In your example, it would test the changed files compared to the last commit in vmarcella/testing and will only run in vmarcella/testing, not in all branches. If a PR already exists, it will update the PR. If it doesn't exist, it creates a new one

File: examples/test-exec-doc.yml
Comment (naman-msft) [reply examples/test-exec-doc.yml:58]: it is needed for the nth commit and is handled by the else block below, where the nth commit and n-1th commit are compared to see the changes and only track those files. for the first commit, it tracks changes based on the last commit of the upstream branch

File: examples/test-exec-doc.yml
Comment (naman-msft) [reply examples/test-exec-doc.yml:95]: done - will add fallback here as well for posterity

File: examples/test-exec-doc.yml
Comment (naman-msft) [reply examples/test-exec-doc.yml]: Let's put it for the future though - it is needed in any case

File: examples/test-exec-doc.yml
Comment (naman-msft) [reply examples/test-exec-doc.yml:164]: good point - updated

File: examples/test-instructions.md
Comment (naman-msft) [reply examples/test-instructions.md:45]: updated

File: examples/test-exec-doc.yml
Comment (vmarcella) [reply examples/test-exec-doc.yml:15]: That's not what I meant, what I was trying to get at is that any branch in the repo that is making changes which include markdown files will automatically have this action run and create a pull request for them, even if they're not done with working on that branch. If that is desired behavior for your example, then that is fine, but this will create PRs whenever someone adds any markdown content to their branch regardless of its executable or if they're done working on it.

File: examples/test-exec-doc.yml
Comment (vmarcella) [reply examples/test-exec-doc.yml:32]: Not quite sure that I follow, could you illustrate an example of how someone would encounter this?

File: examples/test-exec-doc.yml
Comment (vmarcella) [reply examples/test-exec-doc.yml:58]: I was more so referring to the branch checks where you check if the default branch is main or master, but I accidentally highlighted the entire section (Line 36-50). It didn't make sense to me as to why you only check what their default branch is on the first commit when they could have a different default branch on any commit.

File: examples/test-exec-doc.yml
Comment (vmarcella) [reply examples/test-exec-doc.yml:100]: What was the issue with using a pinned version? Because my suggestion is that you just use the pinned version and don't define a fallback. Any version of a python package published to pypi is immutable and it cannot be modified/taken down even if the publisher wanted to.

**Review by vmarcella — CHANGES_REQUESTED** [review]
Continued discussing the open thread but didn't add any further comments. Not relevant to the work itself, but do you think you could actually add another folder called `test-exec-doc` or something along those lines, move the files into there, and then rename `test-instructions.md` to `README.md`? We will have other examples in the future, and this will also allow for your readme to be viewed directly on github as soon as someone enters the example directory.

File: examples/test-exec-doc.yml
Comment (naman-msft) [reply examples/test-exec-doc.yml:32]: as a user, say I created a new branch off of main and didn't make any changes to any markdown files and created a commit, then this will compare the diff with the last commit of main and see there are no unchanged markdown files and not test anything

if I create a new branch off of main and change file1.md, file2.md then it will compare this commit with the last commit of main and see 2 markdown files changed

if I am updating my new branch by changing file1.md, then it will compare the commit to the new branch's last commit and only test file1.md

and if I create a new branch off of something other than main/master (default branches), it tests all the docs there.

File: examples/test-exec-doc.yml
Comment (naman-msft) [reply examples/test-exec-doc.yml:15]: oh good point! let me make it so that the PR creation is an optional step that the customers can toggle on and off in the action itself

File: examples/test-exec-doc.yml
Comment (naman-msft) [reply examples/test-exec-doc.yml:58]: because in those cases, with the nth commit where n>1, there would be an n-1th commit in that very branch. The only time you need to check the default branch is when its the first commit of a new branch off of that branch. Otherwise, that new branch has the previous commit

File: examples/test-exec-doc.yml
Comment (naman-msft) [reply examples/test-exec-doc.yml]: done

**Review by vmarcella — APPROVED** [review]

**Review by vmarcella — APPROVED** [review]

**Pull request merged**
